ncm-dpmlfc: major rewrite to support DAV configuration

jouvin opened the pull request:
This PR is a major rewrite of ncm-dpmlfc to support configuring DPM WebDav protocol with ncm-dpmlfc. To do this, the `ncm-xrootd` rule-based editor has been ported back to DPM to replace the original one. This rule-based editor has been slightly extended but should remain compatible with the `ncm-xrootd` one. The intent is to move it as a new CAF module (it probably represent 1/3 of the configuration module code currently).

Another improvement in `ncm-dpmlfc` is in the schema: rather than specifying the protocol options for each node (whith usually the same options for all nodes), it is now possible to define the protocol options globally that will act as a default for each nodes where those options are not specified.

In addition to this feature extension, there was an extensive rewrite of this configuration module to simplify the code. The old one was based on the assumption that both DPM and LFC could be configured on the same machine, something that has never been actually possible as both product have a (different) daemon listening the same port. This was leading to a very complicated processing of the configuration that has been streamlined. A few 100s of lines have been removed!

As part of this rewrite, unit tests have been added both for some of the rule-based editor features and for the whole `Configure` method with basically all the config files managed by the config module checked. This allowed to uncover quite a number of flaws in the original code and is a pretty good coverage of the whole code. The new version is test friendly: the few `LC::Check` calls still needed have been protected by testing `$CAF::Object::NoAction. When tests are run in verbose mode, the output should produce no `ERROR` message in addition to passing the tests.

This version has been deployed successfully at LAL yesterday and is almost ready for merging. The two remaining minor things to do are:
- Validate pan schema in unit tests
- Update the pod file for the DAV part and the new global options for protocols.

Having it part of 15.12 would be great! Thus the milestone. But I would understand that this is too late...

stdweird commented on ncm-dpmlfc/src/main/pan/components/dpmlfc/config.pan:
```diff
@@ -5,4 +5,16 @@
 unique template components/${project.artifactId}/config;
 
-include { 'components/${project.artifactId}/config-rpm' };
+include { 'components/${project.artifactId}/schema' };
```
no `{}`

stdweird commented on ncm-dpmlfc/src/main/pan/components/dpmlfc/schema.pan:
```diff
@@ -8,6 +8,14 @@ declaration template components/dpmlfc/schema;
+function component_dpmlfc_number_string_valid = {
+  if ( !match(ARGV[0],'^\d+$') ) {
+    error(format("String '%s' is not a valid number",ARGV[0]));
+    return(false);
```
`error` already exits, no need for return

jouvin replied:
I know but I always add it to make it clear to people less aware about pan validation functions... Is it a problem (I have it in many validation functions I wrote in various config modules!)?

stdweird replied:
it's unneeded code and leaving it will confuse more people. i guess it's time to start to cleanup the other code too :smile:

stdweird commented on ncm-dpmlfc/src/main/pan/components/dpmlfc/schema.pan:
```diff
@@ -8,6 +8,14 @@ declaration template components/dpmlfc/schema;
 include { 'quattor/schema' };
 
+function component_dpmlfc_number_string_valid = {
+  if ( !match(ARGV[0],'^\d+$') ) {
```
there's a `is_long` pan builtin i think

jouvin replied:
Unfortunately `is_long` tells you if the variable is a `long` and will return false is you pass it a string... Here the problem is that one resource contains 2 elements, one of them being a string. Thus the other one is a string from pan point of view but has to be interpreted as a number. This validation function ensures that the string is a valid number.

stdweird commented on ncm-dpmlfc/src/main/pan/components/dpmlfc/schema.pan:
```diff
@@ -75,6 +83,20 @@ function component_dpmlfc_node_config_valid = {
 
+# Validation of dav parameters
+function component_dpmlfc_dav_config_valid = {
+   if ( is_defined(SELF['DiskFlags']) ) {
```
`SELF`? if you really do not mean `ARGV[0]`, please document where and/or how to use this function

jouvin replied:
You are right, this is a mistake in fact... I need to fix it. Thanks for for noticing.

stdweird commented on ncm-dpmlfc/src/main/pan/components/dpmlfc/schema.pan:
```diff
@@ -75,6 +83,20 @@ function component_dpmlfc_node_config_valid = {
+function component_dpmlfc_dav_config_valid = {
+   if ( is_defined(SELF['DiskFlags']) ) {
+      foreach (i;flag;SELF['DiskFlags']) {
+        if ( !match(flag,'Write|RemoteCopy|NoAuthn') ) return(false);
```
strict regexp `^(...)$`
also, maybe use error with message inidicating which key is the problematic one

stdweird commented on ncm-dpmlfc/src/main/pan/components/dpmlfc/schema.pan:
```diff
@@ -75,6 +83,20 @@ function component_dpmlfc_node_config_valid = {
+   };
+   if ( is_defined(SELF['NSFlags']) ) {
+      foreach (i;flag;SELF['NSFlags']) {
+        if ( !match(flag,'Write|RemoteCopy|NoAuthn') ) return(false);
```
if this is always the same regexp as above, put the pattern in a local variable.

stdweird commented on ncm-dpmlfc/src/main/pan/components/dpmlfc/schema.pan:
```diff
@@ -125,11 +147,53 @@ type ${project.artifactId}_component_dpns_node_config = {
 
+# DAV related parameters: see zlcgdm-dav.conf for parameter documentation
+type ${project.artifactId}_component_dav_node_config = {
+	"DiskAnonUser" ? string
```
4 space indentation

stdweird commented on ncm-dpmlfc/src/main/pan/components/dpmlfc/schema.pan:
```diff
@@ -125,11 +147,53 @@ type ${project.artifactId}_component_dpns_node_config = {
+	"NSFlags" ? string[]
+	"NSMaxReplicas" ? long
+        "NSRedirectPort" ? long[] with length(SELF) == 2
+	"NSSecureRedirect" ? string with match(SELF,'on|off')
```
no boolean? (at least `^(on|off)$`)

jouvin replied:
The problem is that a boolean will not be translated as on/off by the rule-based editor (it will translate it into yes/no if I'm right). Thus the use of a string.

stdweird commented on ncm-dpmlfc/src/main/pan/components/dpmlfc/schema.pan:
```diff
@@ -125,11 +147,53 @@ type ${project.artifactId}_component_dpns_node_config = {
+	"NSSecureRedirect" ? string with match(SELF,'on|off')
+        "NSServer" ? string[] with length(SELF) == 2 && component_dpmlfc_number_string_valid(SELF[1])
+        "NSTrustedDNs" ? string[]
+        "NSType" ? string with match(SELF,'DPM|LFC')
```
strict regexp

stdweird commented on ncm-dpmlfc/src/main/pan/components/dpmlfc/schema.pan:
```diff
@@ -125,11 +147,53 @@ type ${project.artifactId}_component_dpns_node_config = {
+        "SSLProtocol" ? string[]
+        "SSLSessionCache" ? string
+        "SSLSessionCacheTimeout" ? long
+        "SSLVerifyClient" ? string with match(SELF,'none|optional|require')
```
strict regexp

stdweird commented on ncm-dpmlfc/src/main/pan/components/dpmlfc/schema.pan:
```diff
@@ -125,11 +147,53 @@ type ${project.artifactId}_component_dpns_node_config = {
         "readonly" ? boolean
 };
 
+# DAV related parameters: see zlcgdm-dav.conf for parameter documentation
```
metaconfig supports httpd configuration, why not use that?

jouvin replied:
Because it has to be coordinated with other DPM related config and services. Apache is used but with a specific module that has its own configuration. This is what is managed by this config module.

stdweird commented:
@jouvin the main module is huge. can you split it in one or more modules? in particular the code you want to move to CAF.

* make a new subdir `DPMLFC` in `src/main/perl`
* in that subdir, make perl modules (.e.g `src/main/perl/DPMLFC/mod1.pm`) that start with
```perl
package NCM::Component::DPMLFC::mod1;
...
```
* move functions as-is to that module
 * maybe some import/use and/or module variables can be moved as well (or copied/duplicated)
* in the main module, replace
```perl
use NCM::Component;
use vars qw(@ISA $EC);
@ISA = qw(NCM::Component);
```
with 
```perl
use vars ($EC);
use parent qw(NCM::Component NCM::Component::DPMLFC::mod1);
```
(and add more submodules as you see fit)


* in the `pom.xml`, nothing should be changed (but double check that the rpm picks up the submodules)

this should allow a split in less than 1000 lines modules with very minimal effort; and will allow me to review and comment on the code. in particular, you should not have to touch the code in the methods at all, only imports and class varaibles / constants might be duplicate. also no unittests should be adapted

jouvin commented:
Thanks for the advice. I'll see if I can do something, I concentrated on getting the things working in a first step. If the effort is minimal and it works without tweaking the pom.xml, I'll do it. Else I'd prefer to do the CAF module directly.

stdweird commented:
@jouvin i quickly read the component code, but seems that github doesn't allow me to comment (i can only view the whole file). so splitting the component up will help to review it. (it shouldn't be that that much work, but you have quite a bit of class variables/constants, so those can be more tricky; here's a recent example https://github.com/quattor/CCM/pull/66)

i have some remarks/questions with what are most likely the CAF parts, so this will trigger some discussion if you would move the code to CAF directly (i once also had a similar idea, but never actually tried to come up with code, but it was a different approach). also the code will involve a bit more changes.

however, this shouldn't affect the current component state.

jouvin commented:
I'm reluctant to put any effort into splitting the code into several modules just for the review... I have already spent more time that I wanted on this (useful and too much delayed) rewrite. I'd prefer that we do it as part of the CAF integration and I'm happy to discuss the ideas you have about such a feature. There are probably some other possible implementations but this one proved to work very well for DPM and Xrootd and already covers a lot of use cases. And according to my recent experience, I think it is cleaned enough to be easyly extended.
Before discussing implementation, it may be good to start by discussing the features. I propose to start an issue in CAF with the description of the behaviour (basically the comment in the code before the declaration of the rules. Would it be fine with you?

stdweird commented:
@jouvin sure, i'll see if it can trick github in allowing me to comment on the main module

jouvin commented:
CAF issue opened to discuss the rule-based editor in order to integrate it into CAF: https://github.com/quattor/CAF/issues/123.

jouvin commented:
Af for me, the PR is ready for merging. I understand that reviewing the whole code is quite difficult and is probably not worth the effort as base of the code has been existing for a long time and has been cleaned a lot with the addition of extensive testing. But I'll be happy to address any remark that one of you may have!

jouvin commented:
@stdweird I splitted the code as you suggested to make possible to comment the code in the PR... This is a quick and dirty fix as putting the submodule in @ISA using `use parent` is causing problem with the EXPORT list... The workaround has been to duplicate constant definitions... When moving it as a separate module in CAF, this should be easy to fix. In the meantime, I tried several solutions but all failed!

jouvin commented:
Solution to the export conflict found... Rule-based editor constants no longer duplicated!

stdweird commented on ncm-dpmlfc/src/main/perl/DPMLFC/RuleBasedEditor.pm:
```diff
@@ -0,0 +1,778 @@
+
+use EDG::WP4::CCM::Element;
+
+use Readonly;
```
not used?

jouvin replied:
It is now!

stdweird commented on ncm-dpmlfc/src/main/perl/DPMLFC/RuleBasedEditor.pm:
```diff
@@ -0,0 +1,778 @@
+
+package NCM::Component::DPMLFC::RuleBasedEditor;
+
+use strict;
```
also `use warnings`

jouvin replied:
Done

stdweird commented on ncm-dpmlfc/src/main/perl/DPMLFC/RuleBasedEditor.pm:
```diff
@@ -0,0 +1,778 @@
+# based on the contents of this configuration.
+#
+#######################################################################
+
```
@jouvin i'm going to make a bunch remarks that should be addressed if you would move this code to `CAF`. for the ucrrent PR, i don't care that much

jouvin replied:
No problem! I'll try to fix as many of them as possible during this PR.

stdweird commented on ncm-dpmlfc/src/main/perl/DPMLFC/RuleBasedEditor.pm:
```diff
@@ -0,0 +1,778 @@
+use strict;
+use NCM::Component;
+use vars qw(@ISA $EC);
+@ISA = qw(NCM::Component);
```
`use parent`

stdweird commented on ncm-dpmlfc/src/main/perl/DPMLFC/RuleBasedEditor.pm:
```diff
@@ -0,0 +1,778 @@
+
+# Constants use to format lines in configuration files
+# Keep in sync with ncm-xrootd
+use constant LINE_FORMAT_PARAM => 1;
```
use `Readonly`, to handle exports, see e.g. https://github.com/quattor/configuration-modules-core/blob/master/ncm-systemd/src/main/perl/Systemd/Service.pm#L20 (the trick is to use `our` for exported readonly)

jouvin replied:
I don't know why but I prefer constants for this. With the approach used, it should be easy to manage and don't require any particular variable declaration in the code using it. Currently constants are not exported by default because the trick used with 'use parent' is causing a conflict but after moving it to CAF, I think those constants should be exported by default by `FileEditor` for example.

stdweird commented on ncm-dpmlfc/src/main/perl/DPMLFC/RuleBasedEditor.pm:
```diff
@@ -0,0 +1,778 @@
+# Constants use to format lines in configuration files
+# Keep in sync with ncm-xrootd
+use constant LINE_FORMAT_PARAM => 1;
+use constant LINE_FORMAT_ENVVAR => 2;
```
do these values mean anything, or is this a `enum`-like pattern?

jouvin replied:
This is a enum like pattern...

stdweird commented on ncm-dpmlfc/src/main/perl/DPMLFC/RuleBasedEditor.pm:
```diff
@@ -0,0 +1,778 @@
+use constant LINE_OPT_DEF_ALWAYS_RULES_ONLY => 0;
+
+
+# Define and export the constants used to define rules
```
remove if unused

jouvin replied:
It is used!

jouvin replied:
Sorry, I misinterpreted the comment, you probably speak about the commented lines. In fact this is what I'd like to use when moving to CAF but it doesn't work in the context of a submodule, see comments above.... I may clarify the comments.

stdweird commented on ncm-dpmlfc/src/main/perl/DPMLFC/RuleBasedEditor.pm:
```diff
@@ -0,0 +1,778 @@
+
+#use constant \%rule_constants;
+
+use base 'Exporter';
```
this is a bit dangerous, mixing ISA and base. better `use parent qw(NCM::Component Exporter);`

stdweird commented on ncm-dpmlfc/src/main/perl/DPMLFC/RuleBasedEditor.pm:
```diff
@@ -0,0 +1,778 @@
+
+
+# Backup file extension
+use constant BACKUP_FILE_EXT => ".old";
```
is this some default used in other components? (like filecopy). i think there is somethign similar already used.

jouvin replied:
I am not sure that there is any consistent default across components and I am open to any better suggestion...

stdweird commented on ncm-dpmlfc/src/main/perl/DPMLFC/RuleBasedEditor.pm:
```diff
@@ -0,0 +1,778 @@
+
+
+
+# This function formats an attribute value based on the value format specified.
```
pod style documentation

stdweird commented on ncm-dpmlfc/src/main/perl/DPMLFC/RuleBasedEditor.pm:
```diff
@@ -0,0 +1,778 @@
+
+  my $attr_value = shift;
+  unless ( defined($attr_value) ) {
+    $self->error("$function_name: 'attr_value' argument missing");
```
this sort of errror logging is confusing at best, but not needed at all. the message should at least mention that this is a code error and developers shoudl be contacted, not something wrong with the system or config of dpmlfc.

stdweird commented on ncm-dpmlfc/src/main/perl/DPMLFC/RuleBasedEditor.pm:
```diff
@@ -0,0 +1,778 @@
+#        value_fmt : value format (see LINE_VALUE_xxx constants)
+sub formatAttributeValue () {
+  my $function_name = "formatAttributeValue";
+  my $self = shift;
```
what's wrong with 
```perl
my ($self, $attr_value, $line_fmt, $value_fmt) = @_;
```

jouvin replied:
Probably nothing!

stdweird commented on ncm-dpmlfc/src/main/perl/DPMLFC/RuleBasedEditor.pm:
```diff
@@ -0,0 +1,778 @@
+  my $attr_value = shift;
+  unless ( defined($attr_value) ) {
+    $self->error("$function_name: 'attr_value' argument missing");
+    return 1;
```
1 is `SUCCESS` in CAF, most methods return undef on failure

jouvin replied:
This is probably easy to change. In fact, all methods, except `updateConfigFile` (that should be renamed) are private and should not be used directly.

stdweird commented on ncm-dpmlfc/src/main/perl/DPMLFC/RuleBasedEditor.pm:
```diff
@@ -0,0 +1,778 @@
+#        line_fmt : line format (see LINE_FORMAT_xxx constants)
+#        value_fmt : value format (see LINE_VALUE_xxx constants)
+sub formatAttributeValue () {
+  my $function_name = "formatAttributeValue";
```
4 space indent

jouvin replied:
The whole file is based on 2 space indent... I'm reluctant to change it right now...

stdweird commented on ncm-dpmlfc/src/main/perl/DPMLFC/RuleBasedEditor.pm:
```diff
@@ -0,0 +1,778 @@
+#        attr_value : attribue value
+#        line_fmt : line format (see LINE_FORMAT_xxx constants)
+#        value_fmt : value format (see LINE_VALUE_xxx constants)
+sub formatAttributeValue () {
```
no prototypes, also emthod name on on eline, so 
```perl
sub formatAttributeValue
{
    ...
```

stdweird replied:
i'm also not sure if we use camelcase method names anywhere

jouvin replied:
Is it a problem? I tend to use them, not sure why... Anyway at some point many method name will need to be updated... (at least to start with an `_` for all the internal methods.

stdweird commented on ncm-dpmlfc/src/main/perl/DPMLFC/RuleBasedEditor.pm:
```diff
@@ -0,0 +1,778 @@
+
+  my $formatted_value;
+  if ( $value_fmt == LINE_VALUE_HOST_LIST ) {    
+    # Duplicates may exist as result of a join. Checkt it.
```
typo in `Checkt`

stdweird commented on ncm-dpmlfc/src/main/perl/DPMLFC/RuleBasedEditor.pm:
```diff
@@ -0,0 +1,778 @@
+  $self->debug(2,"$function_name: formatting attribute value >>>$attr_value<<< (line fmt=$line_fmt, value fmt=$value_fmt)");
+
+  my $formatted_value;
+  if ( $value_fmt == LINE_VALUE_HOST_LIST ) {    
```
probably need a lookup table, like https://github.com/quattor/CCM/blob/master/src/main/perl/TextRender.pm#L34
the disadvantage is that you can't log in the methods, but they can be unittested individually, so that's not not such an issue

jouvin replied:
Interesting idea but kept for a later version...

stdweird commented on ncm-dpmlfc/src/main/perl/DPMLFC/RuleBasedEditor.pm:
```diff
@@ -0,0 +1,778 @@
+  if ( $value_fmt == LINE_VALUE_HOST_LIST ) {    
+    # Duplicates may exist as result of a join. Checkt it.
+    my @hosts = split /\s+/, $attr_value;
+    my %hosts;
```
```perl
my %hosts = map { $_ => '' } @hosts;
```

stdweird commented on ncm-dpmlfc/src/main/perl/DPMLFC/RuleBasedEditor.pm:
```diff
@@ -0,0 +1,778 @@
+    # Duplicates may exist as result of a join. Checkt it.
+    my @hosts = split /\s+/, $attr_value;
+    my %hosts;
+    for my $host (@hosts) {
```
use `foreach` (for readability, it's an alias for `for`)

stdweird commented on ncm-dpmlfc/src/main/perl/DPMLFC/RuleBasedEditor.pm:
```diff
@@ -0,0 +1,778 @@
+        $hosts{$host} = "";
+      }
+    }  
+    my $formatted_value="";
```
```perl
my $formatted_value = join(" ", sort keys %hosts);
```

stdweird replied:
also, with `use warnings`, this will cause an error due to redefined `my`

stdweird commented on ncm-dpmlfc/src/main/perl/DPMLFC/RuleBasedEditor.pm:
```diff
@@ -0,0 +1,778 @@
+    $self->debug(1,"Formatted hosts list : >>$formatted_value<<");
+
+  } elsif ( $value_fmt == LINE_VALUE_BOOLEAN ) {
+    if ( $attr_value ) {
```
```perl
$formatted_value = $attr_value ? '"yes"' ? '"no"';
```

(but the boolean values should be readonly)

stdweird commented on ncm-dpmlfc/src/main/perl/DPMLFC/RuleBasedEditor.pm:
```diff
@@ -0,0 +1,778 @@
+
+  } elsif ( $value_fmt == LINE_VALUE_INSTANCE_PARAMS ) {
+    # Instance parameters are described in a nlist
+    if ( exists($attr_value->{logFile}) ) {
```
there's no need for `exists`, is `$attr_value->{logFile}` is undef or empty string, you don't this anyway. so simply
```perl
$formatted_value .= " -l $attr_value->{logFile}" if ($attr_value->{logFile});
```

stdweird commented on ncm-dpmlfc/src/main/perl/DPMLFC/RuleBasedEditor.pm:
```diff
@@ -0,0 +1,778 @@
+    $formatted_value = join " ", @$attr_value;
+
+  } elsif ( $value_fmt == LINE_VALUE_HASH_KEYS ) {
+    $formatted_value = join " ", keys(%$attr_value);
```
`sort keys`

stdweird commented on ncm-dpmlfc/src/main/perl/DPMLFC/RuleBasedEditor.pm:
```diff
@@ -0,0 +1,778 @@
+
+  # Quote value if necessary
+  if ( ($line_fmt == LINE_FORMAT_PARAM) || ($line_fmt == LINE_FORMAT_ENVVAR) ) {
+    if ( $formatted_value =~ /\s+/ ) {
```
also check if it is not quoted yet

stdweird commented on ncm-dpmlfc/src/main/perl/DPMLFC/RuleBasedEditor.pm:
```diff
@@ -0,0 +1,778 @@
+}
+
+
+# This function formats a configuration line using keyword and value,
```
i do not understand why we shouldn't use a TT for this

jouvin replied:
Could you elaborate a little bit more?

stdweird commented on ncm-dpmlfc/src/main/perl/DPMLFC/RuleBasedEditor.pm:
```diff
@@ -0,0 +1,778 @@
+    $config_line =~ s/\s\s+/ /g;
+    $config_line =~ s/\s+$//;
+  } else {
+    $self->error("$function_name: invalid line format ($line_fmt). Internal inconsistency.");
```
maybe use a readonly for the `internal error`/ `Internal inconsistency` type of messages. maybe we need a new reported `$self->deverror` for this type of messages (ie errors that have nothing to do with the system or service, but with the code), but these are different from configuration errors (e.g. when something is missing from the `$config`).
if you think this reasonable, open a issue in CAF for this.

jouvin replied:
Done! https://github.com/quattor/CAF/issues/124

stdweird commented on ncm-dpmlfc/src/main/perl/DPMLFC/RuleBasedEditor.pm:
```diff
@@ -0,0 +1,778 @@
+    return undef;
+  }
+  my $config_value = shift;
+  if ( $config_value ) {
```
should be `if (defined($config_value))` (e.g. `$config_value = 0` etc)

stdweird commented on ncm-dpmlfc/src/main/perl/DPMLFC/RuleBasedEditor.pm:
```diff
@@ -0,0 +1,778 @@
+    $config_value =~ s/\*/\\*/g;
+    $config_value =~ s/\[/\\[/g;
+    $config_value =~ s/\]/\\]/g;
+    $config_value =~ s/\(/[(]/g;
```
why not escape this one?

jouvin replied:
It's an old piece of code, I don't remember!

jouvin replied:
I think I add a problem where escaping `(` was not working and I found this workaround. But I don't see any good reason, was probably a consequence of something else...

stdweird commented on ncm-dpmlfc/src/main/perl/DPMLFC/RuleBasedEditor.pm:
```diff
@@ -0,0 +1,778 @@
+  my $config_value = shift;
+  if ( $config_value ) {
+    $self->debug(2,"$function_name: configuration value '$config_value' will be added to the pattern");
+    $config_value =~ s/\s+/\\s+/g;
```
shouldn't all backslashes be escaped?

stdweird commented on ncm-dpmlfc/src/main/perl/DPMLFC/RuleBasedEditor.pm:
```diff
@@ -0,0 +1,778 @@
+  if ( $config_value ) {
+    $self->debug(2,"$function_name: configuration value '$config_value' will be added to the pattern");
+    $config_value =~ s/\s+/\\s+/g;
+    $config_value =~ s/\-/\\-/g;
```
replace all of below with
```perl
$config_value =~ s/([-+.*[]()^$])/\\1/g; 
```

stdweird commented on ncm-dpmlfc/src/main/perl/DPMLFC/RuleBasedEditor.pm:
```diff
@@ -0,0 +1,778 @@
+    $config_value = "";
+  }
+
+  $config_param =~ s/\s+/\\s+/g;
```
why is this needed? whitespace shouldn't parametrised here.

jouvin replied:
See comment added in the code!

stdweird commented on ncm-dpmlfc/src/main/perl/DPMLFC/RuleBasedEditor.pm:
```diff
@@ -0,0 +1,778 @@
+
+  $config_param =~ s/\s+/\\s+/g;
+  my $config_param_pattern;
+  if ( $line_fmt == LINE_FORMAT_PARAM ) {
```
one advantage of using TT files also here is that at least one can easily verify that the patterns here actually match or are valid for the generated ones in method above.

jouvin replied:
Not sure to understand... I am not familiar enough with TT files. Do you suggest to use a TT file in the rule-based editor or is it a comment on advantages/drawbacks of rule-based editor vs. TT files?

stdweird commented on ncm-dpmlfc/src/main/perl/DPMLFC/RuleBasedEditor.pm:
```diff
@@ -0,0 +1,778 @@
+  $config_param =~ s/\s+/\\s+/g;
+  my $config_param_pattern;
+  if ( $line_fmt == LINE_FORMAT_PARAM ) {
+    $config_param_pattern = "#?\\s*$config_param=".$config_value;
```
the prefix `#?\\s*` is common, maybe put in readonly?

jouvin replied:
Not convinced it improves readibility. The frist characters happen to be identical but we could easily imagine to add the support for a line format where `#` is not the comment prefix...

stdweird commented on ncm-dpmlfc/src/main/perl/DPMLFC/RuleBasedEditor.pm:
```diff
@@ -0,0 +1,778 @@
+    $config_param_pattern = "#?\\s*set\\s+$config_param\\s*=\\s*".$config_value;
+  } elsif ( $line_fmt == LINE_FORMAT_XRDCFG ) {
+    $config_param_pattern = "#?\\s*$config_param";
+    # Avoid adding a withespace requirement if there is no config_value
```
typo in `whitespace`

stdweird commented on ncm-dpmlfc/src/main/perl/DPMLFC/RuleBasedEditor.pm:
```diff
@@ -0,0 +1,778 @@
+
+  $self->debug(1,"$function_name: commenting out lines matching pattern >>>".$config_param_pattern."<<<");
+  # All matching lines must be commented out, except if they are already commented out.
+  # The code used is a customized version of FileEditor::replace() that lacks backreferences.
```
can you open an issue for what is missing exactly?

jouvin replied:
Comment clarified. The problem is that here we want to rewrite the same line commented out but we don't know the current line contents, only a regexp matching it. We would need to express the `newvalue` argument in `replace()` whith some kind of backreference to what was matched by the pattern. This would required the ability to retrieve the group values as we can do in Python with `re.group`. I don't know if this is doable in Perl... if it was, could be a useful extension to `replace()`

stdweird commented on ncm-dpmlfc/src/main/perl/DPMLFC/RuleBasedEditor.pm:
```diff
@@ -0,0 +1,778 @@
+  } else {
+    $self->debug(1, "$function_name: $line_count lines commented out");
+  }
+  $fh->set_contents (join("", @lns));
```
can't this be done with `replace_lines` and a negative lookahead regexp? (`^(?!\s*#)$pattern`)?

jouvin replied:
No, the problem is that `replace')` and `replace_lines()` don't do any parsing on `newvalue` (if I am right), thus any regerence to a group name or similar thing will not be properly translated. Again the code is so simple, I am not sure it is worth to spend efforts extending the current `FileEditor` methods...

stdweird commented on ncm-dpmlfc/src/main/perl/DPMLFC/RuleBasedEditor.pm:
```diff
@@ -0,0 +1,778 @@
+  # The code used is a customized version of FileEditor::replace() that lacks backreferences.
+  my @lns;
+  my $line_count = 0;
+  seek($fh, 0, SEEK_SET);
```
`$fh->seek_begin();`

stdweird commented on ncm-dpmlfc/src/main/perl/DPMLFC/RuleBasedEditor.pm:
```diff
@@ -0,0 +1,778 @@
+
+  my $newline;
+  my $config_param_pattern;
+  $newline = $self->formatConfigLine($config_param,$config_value,$line_fmt);
```
`my $newline = ..` and remove `my $newline` above.
also maybe use `new_line` or just `line` (as a `newline` is something else)

stdweird commented on ncm-dpmlfc/src/main/perl/DPMLFC/RuleBasedEditor.pm:
```diff
@@ -0,0 +1,778 @@
+  } else {
+    $config_param_pattern = $self->buildLinePattern($config_param,$line_fmt);
+  }
+  if ( ($line_fmt == LINE_FORMAT_XRDCFG) && !$multiple ) {
```
move in `else` block above

stdweird commented on ncm-dpmlfc/src/main/perl/DPMLFC/RuleBasedEditor.pm:
```diff
@@ -0,0 +1,778 @@
+                            BEGINNING_OF_FILE,
+                           );
+  
+  # Loop over all config rule entries.
```
move the rule parser in a separate method

stdweird commented on ncm-dpmlfc/src/main/perl/DPMLFC/RuleBasedEditor.pm:
```diff
@@ -0,0 +1,778 @@
+  # written as is, using the line_fmt specified.
+  
+  my $rule_id = 0;
+  foreach my $keyword (sort keys %{$config_rules}) {
```
the `{}` are not needed, `%$config_rules` is ok

stdweird commented on ncm-dpmlfc/src/main/perl/DPMLFC/RuleBasedEditor.pm:
```diff
@@ -0,0 +1,778 @@
+    } else {
+      $condition = "";
+    }
+    $self->debug(1,"$function_name: processing rule ".$rule_id."(variable=>>>".$keyword.
```
no need for the `".$something."`; also you can pass the debug message in multiple arguments, no need to make one long string

stdweird commented on ncm-dpmlfc/src/main/perl/DPMLFC/RuleBasedEditor.pm:
```diff
@@ -0,0 +1,778 @@
+
+    # Check if only rules with ALWAYS conditions must be applied 
+    if ( $parser_options->{always_rules_only} ) {
+      if ( $condition eq "ALWAYS" ) {
```
`ALWAYS` shoudl be a readonly

stdweird commented on ncm-dpmlfc/src/main/perl/DPMLFC/RuleBasedEditor.pm:
```diff
@@ -0,0 +1,778 @@
+                         "condition attribute = '$cond_attribute', negate=$negate");
+      my $cond_satisfied = 1;
+      if ( $cond_attribute ) {
+        # Due to an exists() flaw, testing directly exists($config_options->{$cond_option_set}->{$cond_attribute}) will spring
```
that's not a flaw in exists, it's perl autovivication.
it also means that you can loose the first `exists` and juts write
```perl
my $tmp = $config_options->{$cond_option_set} && exists($config_options->{$cond_option_set}->{$cond_attribute});
```
and use the `$tmp` if both expressions are indeed the same (and should be). it's very hard to compare/review copy/pasted code.

stdweird commented on ncm-dpmlfc/src/main/perl/DPMLFC/RuleBasedEditor.pm:
```diff
@@ -0,0 +1,778 @@
+      if ( $cond_attribute ) {
+        # Due to an exists() flaw, testing directly exists($config_options->{$cond_option_set}->{$cond_attribute}) will spring
+        # into existence $config_options->{$cond_option_set} if it doesn't exist.
+        if ( $negate ) {
```
i don't understand this logic below.

stdweird commented on ncm-dpmlfc/src/main/perl/DPMLFC/RuleBasedEditor.pm:
```diff
@@ -0,0 +1,778 @@
+                                     exists($config_options->{$cond_option_set}->{$cond_attribute});
+        }
+      } elsif ( $cond_option_set ) {
+        if ( $negate ) {
```
this logic contruction is very hard to read.

stdweird commented on ncm-dpmlfc/src/main/perl/DPMLFC/RuleBasedEditor.pm:
```diff
@@ -0,0 +1,778 @@
+          $cond_satisfied = 0 unless exists($config_options->{$cond_option_set});
+        }
+      }
+      # Remove (comment out) configuration line if present and enabled
```
if you comment is this, please don't invert the logic when you wrtie the code. (actually, don't use unless at all :smile: )

stdweird commented on ncm-dpmlfc/src/main/perl/DPMLFC/RuleBasedEditor.pm:
```diff
@@ -0,0 +1,778 @@
+    if ( $attribute ) {
+      for my $option_set (@option_sets) {
+        my $attr_value;
+        if ( $option_set eq "GLOBAL" ) {
```
should be readonly

jouvin commented:
I think I have addressed all comments (except those where the reason for not doing it has been mentioned). Many additional unit tests have been added and a few additional cleanups done. The tests are currently failing on GitHub because `enum` module is used but doesn't seem to be present in the test environment (it is in EPEL from SL5).
Once this is solved, as for me, the PR is ready for merging.

jouvin commented:
Trying to use the new version of the rule-based editor in `ncm-xrootd`, I found a bug in the rule parser (for a use case that I thought covered by the unit tests...). Wait for the fix before merging... Hopefully today.

jouvin commented:
Bug now fixed and the same rule-based editor code is now used in `ncm-xrootd` too (not that the unit tests are only in `ncm-dpmlfc`.

stdweird commented on ncm-dpmlfc/src/main/perl/dpmlfc/RuleBasedEditor.pm:
```diff
@@ -0,0 +1,817 @@
+
+# Constants use to format lines in configuration files
+# Exported constants
+use enum qw(LINE_FORMAT_PARAM=1
```
there's a `=1`, is that for the offset in enum?

jouvin replied:
Yes it must start at 1 for this one (it was the case before, I have not checked if there is a good reason but there is a risk that 0 is used a undefined).

stdweird commented on ncm-dpmlfc/src/main/perl/dpmlfc/RuleBasedEditor.pm:
```diff
@@ -0,0 +1,817 @@
+# ${build-info}
+#
+#
+# This module implement a rule-based editor that is used to modify the content
```
`implement` -> `implements`

stdweird commented on ncm-dpmlfc/src/main/perl/dpmlfc/RuleBasedEditor.pm:
```diff
@@ -0,0 +1,817 @@
+
+
+# Export constants used to build rules
+Readonly my @RULE_CONSTANTS => ('LINE_FORMAT_PARAM',
```
maybe use `qw(LINE_FORMAT_PARAM ...)` no need for commas and quotes

stdweird commented on ncm-dpmlfc/src/main/perl/dpmlfc/RuleBasedEditor.pm:
```diff
@@ -0,0 +1,817 @@
+
+
+# Backup file extension
+use constant BACKUP_FILE_EXT => ".old";
```
readonly

stdweird commented on ncm-dpmlfc/src/main/perl/dpmlfc/RuleBasedEditor.pm:
```diff
@@ -0,0 +1,817 @@
+
+Supported entries for options hash:
+    always_rules_only: if true, apply only rules with ALWAYS condition (D: false)
+    remove_if_undef: if true, remove maatching configuration line is rule condition is not met (D: false)
```
matching

stdweird commented on ncm-dpmlfc/src/main/perl/dpmlfc/RuleBasedEditor.pm:
```diff
@@ -0,0 +1,817 @@
+=cut
+
+sub updateConfigFile {
+  my $function_name = "updateConfigFile";
```
4 space indent

jouvin replied:
Same as before, I'll change for 4 space ident later, when everything is clean...

stdweird commented on ncm-dpmlfc/src/main/perl/dpmlfc/RuleBasedEditor.pm:
```diff
@@ -0,0 +1,817 @@
+
+  # Quote value if necessary
+  if ( ($line_fmt == LINE_FORMAT_PARAM) || ($line_fmt == LINE_FORMAT_ENVVAR) ) {
+    if ( (($formatted_value =~ /\s+/) && ($formatted_value !~ /^".*"$/)) ||
```
what with single quotes?

stdweird commented on ncm-dpmlfc/src/main/perl/dpmlfc/RuleBasedEditor.pm:
```diff
@@ -0,0 +1,817 @@
+  if ( defined($config_value ) ) {
+    $self->debug(2,"$function_name: configuration value '$config_value' will be added to the pattern");
+    $config_value =~ s/\\/\\\\/g;
+    $config_value =~ s/([\-\+\?\.\*\[\]()\^\$])/\\$1/g;
```
use `\g1` as backref

jouvin replied:
For me (and based on my test), '\g1` can be used only inside the regexp, no on the substitution side....

stdweird replied:
you're right, i was confused a bit

stdweird commented on ncm-dpmlfc/src/main/perl/dpmlfc/RuleBasedEditor.pm:
```diff
@@ -0,0 +1,817 @@
+
+  $self->debug(1,"$function_name: commenting out lines matching pattern >>>".$config_param_pattern."<<<");
+  # All matching lines must be commented out, except if they are already commented out.
+  # The code used is a customized version of FileEditor::replace() that lacks support for backreferences
```
open an issue with an example, i still don't fully see the case.
but shouldn't be too hard to add and if it's useful, i see no point why not.

jouvin replied:
Done! https://github.com/quattor/CAF/issues/125.

stdweird commented on ncm-dpmlfc/src/main/perl/dpmlfc/RuleBasedEditor.pm:
```diff
@@ -0,0 +1,817 @@
+
+=item _updateConfigLine
+
+This function do the actual update of a configuration line after doing the final
```
`do` -> `does`

stdweird commented on ncm-dpmlfc/src/main/perl/dpmlfc/RuleBasedEditor.pm:
```diff
@@ -0,0 +1,817 @@
+    }
+
+    # Check if rule condition is met if one is defined
+    unless ( $condition eq "" ) {
```
try to avoid unless, this is just `if ($condition ne '') {`

stdweird commented on ncm-dpmlfc/src/main/perl/dpmlfc/RuleBasedEditor.pm:
```diff
@@ -0,0 +1,817 @@
+
+=cut
+
+sub _parse_rules {
```
the rule parser should only parse rules and return bunch of boolean / values, like `$condition, `$negate`,... 
mixing the actual configuration changes makes this hard to read (and a very long function)

jouvin replied:
@stdweird I understand your point but the problem is how do you return a bunch of value from a Perl function (I know how to do it in Python!). You basically need to return if the rule has to be applied or not and the parameter used to edit the line after the rule is parsed. That means one boolean and one hash. Do you consider that passing the hash as an argument and having it modified by the function is a good practice?

stdweird commented on ncm-dpmlfc/src/main/perl/dpmlfc/RuleBasedEditor.pm:
```diff
@@ -0,0 +1,817 @@
+        # $config_options->{$cond_option_set} into existence if it doesn't exist.
+        my $cond_true = exists($config_options->{$cond_option_set}) && 
+                        exists($config_options->{$cond_option_set}->{$cond_attribute});          
+        if ( $negate ) {
```
maybe rewrite this to 
```perl
$cond_true = ! $cond_true if $negate
$cond_satisfied = 0 if $cond_true;
```
(altough this changes `$cond_true`, don't know if it is used below)

stdweird commented on ncm-dpmlfc/src/main/perl/dpmlfc/RuleBasedEditor.pm:
```diff
@@ -0,0 +1,817 @@
+      if ( $cond_attribute ) {
+        # Due to Perl autovivification, testing directly exists($config_options->{$cond_option_set}->{$cond_attribute}) will spring
+        # $config_options->{$cond_option_set} into existence if it doesn't exist.
+        my $cond_true = exists($config_options->{$cond_option_set}) && 
```
first exists is not needed

stdweird commented on ncm-dpmlfc/src/main/perl/dpmlfc/RuleBasedEditor.pm:
```diff
@@ -0,0 +1,817 @@
+            $attribute_present = 0;
+          }
+        } else {
+          # Due to an exists() flaw, testing directly exists($config_options->{$cond_option_set}->{$cond_attribute}) will spring
```
replace with `# See autovivication comment above`

stdweird commented on ncm-dpmlfc/src/main/perl/dpmlfc/RuleBasedEditor.pm:
```diff
@@ -0,0 +1,817 @@
+        } else {
+          # Due to an exists() flaw, testing directly exists($config_options->{$cond_option_set}->{$cond_attribute}) will spring
+          # into existence $config_options->{$cond_option_set} if it doesn't exist.
+          if ( exists($config_options->{$option_set}) && exists($config_options->{$option_set}->{$attribute}) ) {
```
first exists is not needed

jouvin commented:
@stdweird what about `perl-enum` not found during executiong of unit tests? Is it enough to list it as a dependency in the `pom` file?

stdweird commented:
test this please

stdweird commented:
@jouvin i added enum, tests pass again. (i forgot to do that yesterday, apologies for the delay)

stdweird commented on ncm-dpmlfc/src/main/perl/dpmlfc/RuleBasedEditor.pm:
```diff
@@ -0,0 +1,816 @@
+
+=cut
+
+sub _parse_rules {
```
the previous conversation was lost

> I understand your point but the problem is how do you return a bunch of value from a Perl function (I know how to do it in Python!). 
i would convert the curent variables in a hashref, and return the hashref

e.g. 
```perl
sub _parse_one_rule
{
    ...
    my $res = {};
    $res->{value_fmt} = $value_fmt;
    ...
    return $res;
}
```
(i'm not sure i understand your question though)
> You basically need to return if the rule has to be applied or not and the parameter used to edit the line after the rule is parsed. That means one boolean and one hash. Do you consider that passing the hash as an argument and having it modified by the function is a good practice?
i am talking about a rule parser, i.e. something that converts
```
[condition->]attribute:option_set[,option_set,...];line_fmt
```
in 
```
my $rule = {
    condition => x,
    attrinute => y,
    option_set => [option1, ...],
    line_fmt =>,
}
```
in particular, this method should retrun undef on failure and log error in case the rule is invalid, so it can be used e.g. for unittestting the rules. in particular, all rules should be valid/parsable before a single one is applied.
right now, you apply the rules one by one, and in case there is an issue, you continue, instead of doing a `$fh->cancel` and return undef/stop.

jouvin commented:
With the last commit, the suggestion to have a distinct method for parsing rules has been implemented. Updated in both `ncm-dpmlfc` and `ncm-xrootd`. 2 improvements are left for a later version:
- cancel file modification if an error occurs applying of the rules. I agree that this should be the default behaviour but I'd like to implement the option to just ignore errors (apply as many mods as possible) to keep the current behaviour possible.
- Preserve initial whitespace when modifying a line. This should be possible with the customized `replace()` used and the required feature may be provided by `CAF::FileEditor` in the future (https://github.com/quattor/CAF/issues/123).

As for me, this PR is ready for merging.

jouvin commented:
BTW, deployed successfully at LAL.

jouvin commented:
@jrha I'd really like to see this one going into 15.12, please... It has been under a very detailed review by @stdweird  and is in prod at LAL. It will bring only advantages on the previous version!

jouvin commented:
A bug was just uncovered at LAL in `shift.conf` management. I'm fixing it asap. Unfortunately this file has a rule with a specific feature and was not enough unit tested, conversely to other files... Another proof that we should never minimize the effort on unit tests!

jouvin commented:
Pb fixed, new version deployed successfully at LAL, missing coverage in unit tests fixed.

@jrha is there any change to have it included in 15.12, I saw that RC2 is out already, I know this is late but there are some important fixes for people using Quattor to manage DPM...

jrha commented:
I spend a while thinking about this, but in the end decided that given it only touches one grid component which is known to have issues, it was better off merged.

jouvin commented:
Thanks!
